Add PEST_HP Lambda Functionality to PEST++, modify clamping behavior

mmorphew opened the pull request:
(n.b. There are some issues with the windows tests I am working on resolving. There is also some code from usgs/develop that is not getting properly pulled into my downstream branch. I may add it manually if the pull request still isn't catching the discrepancy. Finally, I need to do a bit more testing to ensure that when running in parallel, the pest HP lambdas are behaving as intended. Some of these issues are easier to identify and fix in a draft PR, hence why I'm creating this PR a bit early.)


A longtime request from colleagues has been for a simple "Let me do lambdas like how PEST_HP does it" setting for pestpp-glm. The _++glm_hp_lambdas()_ control file option has been added to do exactly that. When set to _true_, this overrides GLM's current lambda and lambda scale vector setup. Instead, lambdas and lambda scaling factors will be drawn according to how PEST_HP does it. In a nutshell, a central lambda value is picked as a function of "PHI / NUM_NONZEROWEIGHT_OBS". Higher and lower values are then fanned out from that central value. If running through Panther, all Panther agents will be used for the upgrade process. This change has required me to expose one of the panther manager's functions `get_agent_stats` as public. `rm_yamr` has also been added to pestpp-common's CMake list. Other than that, there shouldn't be any other major changes to the code outside of my usual Pest.cpp and SVDSolver.cpp space for this kind of work.


Finally, I have modified my older clamping code to be a bit more conservative. It will now clamp to the 5/95 percentile of a parameter's range, rather than the absolute bounds. I believe Jeremy's latest glm hotfix has already corrected the bounds issues that I have been noticing, but this change adds another layer of protection.


The docs have also been updated to reflect the new glm_hp_lambdas setting.

Commit:
Update clamping code to only clamp to 5 or 95% of range rather than clamp to the endpoints

Commit:
Merge branch 'master' of github.com:mmorphew/pestpp

mmorphew committed:
Merge branch 'usgs:master' into master

Commit:
Update parchange limit test to account for changes to clamping

Commit:
Merge branch 'master' of github.com:mmorphew/pestpp

Commit:
Add HP-style lambdas to GLM

Commit:
Attempt to fix pathing

Commit:
Edit the right file this time

Commit:
Trying new pathing

Commit:
Trying robust fix

Commit:
Trying robust fix 2

Commit:
Edit the right file 2

Commit:
Tighten up HP lambdas, port to SVDA solver

Commit:
Rename HP lambdas variable, push changes to SVDA solver, minor tweaks so that the option better aligns with PEST_HP behavior

Commit:
Add yamr to target link libraries

Commit:
Correct lib name

Commit:
Add second bounds check for clamping code

Commit:
Merge branch 'develop' of https://github.com/usgs/pestpp

Commit:
Fix merge

Commit:
One more user-controlled lambda option

Commit:
Trying to fix windows build

Commit:
Add some headers to SVDA solver

jtwhite79 commented:
The first place I saw the phi-based initial lambda was in the Chen and Oliver 2013 paper (in Algorithm 1) but thats just for setting the initial lambda. IES uses that approach, but then also does some adaptive lambda stuff like increasing lambda if the iteration does not meet the phi reduction thresholds...besides the lambda, the way the truncated SVD is implemented can also make a large difference in the quality of the upgrades so maybe its worth looking at that as well...

Commit:
Continue debugging windows, try fixing long run times on github actions

Commit:
Fix upstream merge

jtwhite79 committed:
mac build

jtwhite79 committed:
Merge remote-tracking branch 'origin'

Commit:
Resolve merge conflicts, need to double check SVDASolver later

Commit:
Attempt to fix environment runner

Commit:
Failed a spot check on why SVDASolver was so different

Commit:
Remove attempted fix

Commit:
Merge branch 'master' of https://github.com/mmorphew/pestpp

Commit:
Debugging discrepency in GLM vs. HP performance when using HP lambdas

Commit:
Debugging discrepency in GLM vs. HP performance when using HP lambdas

Commit:
Add HP normal matrix scaling; Eureka!

Commit:
Reset bound tightening

Commit:
Merge branch 'master' of github.com:mmorphew/pestpp

Commit:
Undo merge issue

Commit:
Correct some comments

Commit:
Cleaning up user manual a bit

mmorphew commented:
Reopening this. Apologies for the extended hiatus on this PR - I haven't had time to work on this. 

I was finally able to hit rough parity between GLM and HP performance on a very simple model by adding another option ("HP") for GLM normal mode matrix scaling [glm_normal_form(HP)]. Unfortunately, as far as I can tell, this is a section of the PEST/PEST_HP codebase that never quite made it into a manual, so I am not sure of the exact justifications regarding exactly how this scaling is performed. I will do some searching for any papers that reference this code, similar to how Chen and Oliver referencing using phi as a basis for setting initial lambda. 

At the risk of stating the obvious, perfect parity between GLM and HP is impossible, nor is it desirable. These are software written in two completely different languages. And while PEST_HP is hardened, it's not infallible in terms of its implementation of optimizing L-M. My goal with this and the clamping PR is/was instead to identify the main blockers arresting PEST++GLM's performance and adoption, and then modify those blockers so that GLM performs as reasonably as expected. And as previously discussed, perhaps some of these findings have some kind of portability to IES.

Commit:
Attempting to safeguard against zero sen values

Commit:
More checks against bad factorization

Commit:
Rethinking bounds checking for clamping to accomodate negative parameter values

mmorphew commented:
Apologies for continually closing/reopening this, I'm going to pull this down temporarily while I fix some persistent issues with parameter bounds being violated in some real world use cases.

jtwhite79 commented:
all good.  Appreciate your efforts.  Just FYI: pestpp-glm got some refactoring lately, so make sure you either merge with upstream or start fresh...

Commit:
Attempting another bounds fix

mmorphew committed:
Merge branch 'usgs:master' into master

Commit:
Trying to match transformation logic in the other branch

Commit:
Merge branch 'master' of github.com:mmorphew/pestpp

Commit:
Another attempt

Commit:
Crazy idea where we skip the frozen par check upon upgrades and just clamp everything

Commit:
Add checks

Commit:
Why is this passing locally

Commit:
What if we moved clamping to the freezing code and reset enforce_par_limits to what it was before?

Commit:
Fix final check

Commit:
Messing about with lambda scale checks and parameter transformations

Commit:
avoiding bound check for initial freeze check to avoid removing all active parameters from play

Commit:
Add enforce par limit to end of check_bnd_par since we're now clamping there

Commit:
Reset tolerance

Commit:
Check par limits individually for frozen parameter now; update comments and .rec output

mmorphew commented:
Thanks Jeremy, I did a pull from upstream yesterday and will keep an eye on any fixes that would affect this PR being cleanly merged.

I plan to reopen this soon now that tests are passing again, but I want to do some more in-depth testing on some larger models before taking this out of draft/closed status.

In short, I went back through the PEST_HP source code and realized I misread what it was doing pretty badly in terms of clamping. (A bit of a classic case of "seeing what you want to see.") As a result, I've reset `enforce_par_limits` back to where it was before my first PR. The clamping code instead has been moved into `check_bnd_par`, where parameters offending their bounds during an upgrade are clamped either to the boundary or their change limit if the boundary is too far to reach. 

`check_bnd_par` is called by two functions at various points throughout the iteration upgrade process. When it is used to test and find parameters that should be frozen during `test_upgrade_to_find_freeze_pars`, only parameters at their bounds and heading out are checked (`include_bounds` set to false). This is identical to `check_bnd_par`'s current behvaior when it is called during `test_upgrade_to_find_freeze_pars` and prevents the entire upgrade set from being unintentionally frozen. However, when `check_bnd_par` is called during `calc_upgrade_vec`, `include_bounds` has been switched from false to true, so all bounds violators are checked for and clamped rather than set to their original values. Boundary violators are checked against their parameter change limits individually to avoid universal shrinkage becoming too high for the entire set of to-be frozen parameters. 

I still need to test this further, but my hope this is that is a much, much cleaner implementation than the previous attempt at this behavior in `enforce_par_limits`, while preventing GLM's performance from getting hamstrung by a parameter that wants to move leaps and bounds further than the other parameters.

Commit:
Trying to ensure parameters at bounds and going out of bounds don't get their limits checked

Commit:
Trying to tighten up not passing frozen parameters during scaled lambda bounds enforcement checks

Commit:
Make bounds tolerances equal between pest.cpp and svdsolver.cpp, add a hard check to enforce_par_limit to check for parameters that were/are at bounds

mmorphew commented:
Checks are passing on my fork, I think the Windows development environment on this side just had a tiny hiccup when pulling from GitHub.


I've done some testing on a larger groundwater model which revealed some nasty edge cases that unfortunately are hard to test for with unit testing. Think situation like "this parameter get to just outside the tolerance of one of its bounds during iteration 5 and then gets scaled by a lambda scaler in such a way that `enforce_par_limit` throws an error." 


I wanted to avoid editing `Pest.cpp` due to it touching multiple executables in the PEST++, but I eventually caved and added a check to `enforce_par_limits` to skip bounds checking if both the last value and the current value are within a tolerance of one of the parameter bounds. This avoids calculating a scale factor greater than 1 (usually infinity). Other than that, most of the other additions from this PR are within `SVDSolver.cpp` and should only be touching GLM. I've added more robust checks to `check_bnd_par` to make sure that parameters aren't being moved to their bounds and frozen unintentionally, and I've added some clunky for loops to try to make the `enforce_par_limits` check during the lambda vector scaling in `iteration_upgrd` less likely to send a frozen parameter into that function.


Again, my apologies for this PR winding up being a bit of a junior development PR with how many changes (and reversed changes) it wound up encompassing. I know that's less than ideal and that it'll take a while to review. As always, I appreciate your time and letting me tinker with pestpp-glm.

mmorphew committed:
Update glm tests

jtwhite79 committed:
Merge branch 'develop' into master
